File: tests/rosetta/transpiler/Clojure/bulls-and-cows-player.clj
(ns main (:refer-clojure :exclude [indexOf fields makePatterns main]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare indexOf fields makePatterns main)

(declare fields_ch fields_cur fields_i fields_words indexOf_i main_b main_bulls main_c main_cg main_cows main_cp main_guess main_i main_idx main_line main_pat main_patterns main_toks makePatterns_digits makePatterns_i makePatterns_j makePatterns_k makePatterns_l makePatterns_pats next_v)

(defn indexOf [indexOf_s indexOf_ch]
  (try (do (def indexOf_i 0) (while (< indexOf_i (count indexOf_s)) (do (when (= (subs indexOf_s indexOf_i (+ indexOf_i 1)) indexOf_ch) (throw (ex-info "return" {:v indexOf_i}))) (def indexOf_i (+ indexOf_i 1)))) (throw (ex-info "return" {:v (- 1)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn fields [fields_s]
  (try (do (def fields_words []) (def fields_cur "") (def fields_i 0) (while (< fields_i (count fields_s)) (do (def fields_ch (subs fields_s fields_i (+ fields_i 1))) (if (or (or (= fields_ch " ") (= fields_ch "\t")) (= fields_ch "\n")) (when (> (count fields_cur) 0) (do (def fields_words (conj fields_words fields_cur)) (def fields_cur ""))) (def fields_cur (str fields_cur fields_ch))) (def fields_i (+ fields_i 1)))) (when (> (count fields_cur) 0) (def fields_words (conj fields_words fields_cur))) (throw (ex-info "return" {:v fields_words}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn makePatterns []
  (try (do (def makePatterns_digits ["1" "2" "3" "4" "5" "6" "7" "8" "9"]) (def makePatterns_pats []) (def makePatterns_i 0) (while (< makePatterns_i (count makePatterns_digits)) (do (def makePatterns_j 0) (while (< makePatterns_j (count makePatterns_digits)) (do (when (not= makePatterns_j makePatterns_i) (do (def makePatterns_k 0) (while (< makePatterns_k (count makePatterns_digits)) (do (when (and (not= makePatterns_k makePatterns_i) (not= makePatterns_k makePatterns_j)) (do (def makePatterns_l 0) (while (< makePatterns_l (count makePatterns_digits)) (do (when (and (and (not= makePatterns_l makePatterns_i) (not= makePatterns_l makePatterns_j)) (not= makePatterns_l makePatterns_k)) (def makePatterns_pats (conj makePatterns_pats (str (str (str (nth makePatterns_digits makePatterns_i) (nth makePatterns_digits makePatterns_j)) (nth makePatterns_digits makePatterns_k)) (nth makePatterns_digits makePatterns_l))))) (def makePatterns_l (+ makePatterns_l 1)))))) (def makePatterns_k (+ makePatterns_k 1)))))) (def makePatterns_j (+ makePatterns_j 1)))) (def makePatterns_i (+ makePatterns_i 1)))) (throw (ex-info "return" {:v makePatterns_pats}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn main []
  (try (do (println (str (str (str (str (str "Cows and bulls/player\n" "You think of four digit number of unique digits in the range 1 to 9.\n") "I guess.  You score my guess:\n") "    A correct digit but not in the correct place is a cow.\n") "    A correct digit in the correct place is a bull.\n") "You give my score as two numbers separated with a space.")) (def main_patterns (makePatterns)) (while true (do (when (= (count main_patterns) 0) (do (println "Oops, check scoring.") (throw (ex-info "return" {:v nil})))) (def main_guess (nth main_patterns 0)) (def main_patterns (subvec main_patterns 1 (count main_patterns))) (def main_cows 0) (def main_bulls 0) (loop [while_flag_1 true] (when (and while_flag_1 true) (do (println (str (str "My guess: " main_guess) ".  Score? (c b) ")) (def main_line (read-line)) (def main_toks (fields main_line)) (cond (= (count main_toks) 2) (do (def main_c (Integer/parseInt (nth main_toks 0))) (def main_b (Integer/parseInt (nth main_toks 1))) (when (and (and (and (and (>= main_c 0) (<= main_c 4)) (>= main_b 0)) (<= main_b 4)) (<= (+ main_c main_b) 4)) (do (def main_cows main_c) (def main_bulls main_b) (recur false)))) :else (do (println "Score guess as two numbers: cows bulls") (recur while_flag_1)))))) (when (= main_bulls 4) (do (println "I did it. :)") (throw (ex-info "return" {:v nil})))) (def next_v []) (def main_idx 0) (while (< main_idx (count main_patterns)) (do (def main_pat (nth main_patterns main_idx)) (def main_c 0) (def main_b 0) (def main_i 0) (while (< main_i 4) (do (def main_cg (subs main_guess main_i (+ main_i 1))) (def main_cp (subs main_pat main_i (+ main_i 1))) (if (= main_cg main_cp) (def main_b (+ main_b 1)) (when (>= (indexOf main_pat main_cg) 0) (def main_c (+ main_c 1)))) (def main_i (+ main_i 1)))) (when (and (= main_c main_cows) (= main_b main_bulls)) (def next_v (conj next_v main_pat))) (def main_idx (+ main_idx 1)))) (def main_patterns next_v)))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/rosetta/transpiler/FS/bulls-and-cows-player.fs
// Generated 2025-07-27 15:57 +0700

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

open System

let rec indexOf (s: string) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length s) do
            if (_substring s i (i + 1)) = ch then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
and fields (s: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    try
        let mutable words: string array = [||]
        let mutable cur: string = ""
        let mutable i: int = 0
        while i < (String.length s) do
            let ch: string = _substring s i (i + 1)
            if ((ch = " ") || (ch = "\t")) || (ch = "\n") then
                if (String.length cur) > 0 then
                    words <- unbox<string array> (Array.append words [|cur|])
                    cur <- ""
            else
                cur <- cur + ch
            i <- i + 1
        if (String.length cur) > 0 then
            words <- unbox<string array> (Array.append words [|cur|])
        __ret <- unbox<string array> words
        raise Return
        __ret
    with
        | Return -> __ret
and makePatterns () =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    try
        let digits: string array = [|"1"; "2"; "3"; "4"; "5"; "6"; "7"; "8"; "9"|]
        let mutable pats: string array = [||]
        let mutable i: int = 0
        while i < (unbox<int> (Array.length digits)) do
            let mutable j: int = 0
            while j < (unbox<int> (Array.length digits)) do
                if j <> i then
                    let mutable k: int = 0
                    while k < (unbox<int> (Array.length digits)) do
                        if (k <> i) && (k <> j) then
                            let mutable l: int = 0
                            while l < (unbox<int> (Array.length digits)) do
                                if ((l <> i) && (l <> j)) && (l <> k) then
                                    pats <- unbox<string array> (Array.append pats [|(((digits.[i]) + (digits.[j])) + (digits.[k])) + (digits.[l])|])
                                l <- l + 1
                        k <- k + 1
                j <- j + 1
            i <- i + 1
        __ret <- unbox<string array> pats
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        printfn "%s" ((((("Cows and bulls/player\n" + "You think of four digit number of unique digits in the range 1 to 9.\n") + "I guess.  You score my guess:\n") + "    A correct digit but not in the correct place is a cow.\n") + "    A correct digit in the correct place is a bull.\n") + "You give my score as two numbers separated with a space.")
        let mutable patterns: string array = makePatterns()
        try
            while true do
                if (unbox<int> (Array.length patterns)) = 0 then
                    printfn "%s" "Oops, check scoring."
                    __ret <- ()
                    raise Return
                let guess: string = patterns.[0]
                patterns <- unbox<string array> (Array.sub patterns 1 ((unbox<int> (Array.length patterns)) - 1))
                let mutable cows: int = 0
                let mutable bulls: int = 0
                try
                    while true do
                        printfn "%s" (("My guess: " + guess) + ".  Score? (c b) ")
                        let line: string = System.Console.ReadLine()
                        let toks: string array = fields line
                        if (unbox<int> (Array.length toks)) = 2 then
                            let c: int = int (toks.[0])
                            let b: int = int (toks.[1])
                            if ((((c >= 0) && (c <= 4)) && (b >= 0)) && (b <= 4)) && ((c + b) <= 4) then
                                cows <- c
                                bulls <- b
                                raise Break
                        printfn "%s" "Score guess as two numbers: cows bulls"
                with
                | Break -> ()
                | Continue -> ()
                if bulls = 4 then
                    printfn "%s" "I did it. :)"
                    __ret <- ()
                    raise Return
                let mutable next: string array = [||]
                let mutable idx: int = 0
                while idx < (unbox<int> (Array.length patterns)) do
                    let pat: string = patterns.[idx]
                    let mutable c: int = 0
                    let mutable b: int = 0
                    let mutable i: int = 0
                    while i < 4 do
                        let cg: string = _substring guess i (i + 1)
                        let cp: string = _substring pat i (i + 1)
                        if cg = cp then
                            b <- b + 1
                        else
                            if (unbox<int> (indexOf pat cg)) >= 0 then
                                c <- c + 1
                        i <- i + 1
                    if (c = cows) && (b = bulls) then
                        next <- unbox<string array> (Array.append next [|pat|])
                    idx <- idx + 1
                patterns <- unbox<string array> next
        with
        | Break -> ()
        | Continue -> ()
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
